### ISSUE TITLE
Excessive line breaks for a short comparison expression

### ISSUE THREAD

**Issue by atomiks** [prettier/prettier#3621]
**Prettier 1.9.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEAHAhgGwEsoATAJwgFs5MBeTDMugPk2AB0pNNIoMG6DGGQB0MCABkIAdzhkAwrjRwAFAEoRZONny4wqgPQBtduwDqAfQC6BgOYAaTAHInagNyduWmAFcyXNDpaejQRNB1CGBUXDS0ANzlldREAKwhiaNcPKABfNxB7EAhsGEJ0ZFBcMgppAAUqhDRkEFw49JICloxkADMCZUKAIzI9AGs4GABlPDBiW2RhHzhCkggwXv7lkGJlMhhakdtKXA38AZAUtAAPACERsHGp3GoJYjhT84gfGGxvgCYPls8GRds1BrhBgBPfDQTrYMjEGBmQgkGAAC2QAA4AAyFeEQZRmEbYZrwuC7BKdLQARx8hC0B1wRxOSD6Zy2ykohAWZCWhTQc3wcAAij4IPBAYUYBDkaiMUg-lKRoQiFBbPIqMdmlBoO9Cj5lAAVCFNVmbHI5IA)

**Input:**
```jsx
const palindrome = str => {
  const s = str.toLowerCase().replace(/[\W_]/g, '');
  return s === s.split('').reverse().join('');
};
```

**Output:**
```jsx
const palindrome = str => {
  const s = str.toLowerCase().replace(/[\W_]/g, "");
  return (
    s ===
    s
      .split("")
      .reverse()
      .join("")
  );
};

```

**Expected behavior:**
The line breaks probably shouldn't occur in this case. It's excessive and doesn't look very nice.

**Comment by lydell** [comment]
This is definitely not very good looking ;)

It’s really hard to come up with a rule for this, though.

**Comment by j-f1** [comment]
Maybe we should only break if one of the functions is passed a function (or the chain goes past `maxWidth`), meaning that Promises would continue to break.

**Comment by a-heryani** [comment]
I am seeing similar behaviour, although not as bad 

![prettier](https://user-images.githubusercontent.com/26278881/34610274-85706e8c-f274-11e7-976a-d7293d28ce94.gif)

**Comment by Tanoemon** [comment]
How about this jsx?

**Input:**
```
<ol ref={ol => { this.ol = ol}}>
</ol>
```

**Output:**
```
<ol
  ref={ol => {
    this.ol = ol
  }}
>
</ol>
```

I prefer one line function because readable markdown readability is more important than function readability in this case.

**Comment by j-f1** [comment]
Can you open a separate issue for that? It appears unrelated.

**Comment by Sheile** [comment]
Following example is a problem for our test code.

**Prettier 1.11.1**

[Playground](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEcAeAHOYYAoxwA2hAzgHQCGeAjAJRkBOCAJnA7vfGnrZxAEJxcAckIBLKAGsABFxjVhtADpR0WHPiKlKeAEz0mUVu07oefQSPFTZZ3YpVrseAsXJVcAZgMs2HMnIcFkKiEjJyng6qmM6abjq4ACw+Rn6m3EEBAiHW4WaJUSAANCAQGDBi0CTIoBQMDBAA7gAKdQjVKBQAbhBizMUgFCQwyABmFKRwJQBGDBRgknAwAMoY8xIA5sgwDACuUyDMEGBjEyQHEucMMM1zGwC2FKeTJQBWJGj8cwtLyxT3cAAMhI4M9ziUILsYBgobowQc1gwrsgQNMKNMAJ6EaADDAMCQwADqfRgAAtkAAOAAMJTxEHOhLmGBReLgVy6oJKTAAjrsxExbhQHk8kOMXiBzvcxNs9gcSJtCHAAIq7CDweElGDo4nMMnIXSauZiawbADCEHujxRUGgnJAu3OABV0R0xecAL7uoA)

**Input:**
```
expect(cells.at(1).render().text()).toBe('link text1')
expect(cells.at(2).render().text()).toBe('link text2')
expect(cells.at(3).render().text()).toBe('link text3')
expect(cells.at(4).render().text()).toBe('link text4')
```

**Output:**
```
expect(
  cells
    .at(1)
    .render()
    .text()
).toBe("link text1");
expect(
  cells
    .at(2)
    .render()
    .text()
).toBe("link text2");
...
```

**j-f1 wrote:**
> Maybe we should only break if one of the functions is passed a function (or the chain goes past `maxWidth`)

Nested functions with `expect` are very popular in specs.

I'll disable this rule in specs if prettier can configured. but I already understand the option philosophy in prettier.
I hope that only break if the chain goes past `maxWidth`.

**Appendix**
Above example can convert to other logic that expect equality with array, 
FYI, I thought other example like below.
```
expect(cells.at(1).text()).toBe('plain text1')
expect(cells.at(2).text()).toBe('plain text2')
expect(cells.at(3).render().text()).toBe('link text1')
expect(cells.at(4).render().text()).toBe('link text2')
```

**Comment by selvagsz** [comment]
Couldn't incorporate prettier in our codebases bcoz of this method chaining issue. Respecting the printWidth would be best option

**Comment by suchipi** [comment]
Ah.

Maybe we should go back on that "3 or more chained calls" part? A few people have raised issues about this.

That said, we should dig up more context first before making a change...

Thanks for finding that.

**Comment by duailibe** [comment]
The discussion about this in #1282 and #1565. The PR that introduced the behavior in #2673.

**Comment by suchipi** [comment]
Thanks @duailibe!

We could probably adjust this slightly with a new heuristic to address the edge cases we've seen, but I don't have a proposal at this time.

The most common place I see this raised is with enzyme, so that could impact how we implement the heuristic (if we do so).

**Comment by lehni** [comment]
I would love to see this addressed. It is the main reason why I am still refraining from using prettier in my projects.

The  "3 or more chained calls" rule seems arbitrary to me. I would suggest that there should be an option that allows changing this numeric threshold, and optionally the option could be set to a string that tells it to respect the `printWidth`, e.g.:
```json
{
  "chainedCallThreshold": 4
}
```
OR
```json
{
  "printWidth": 120,
  "chainedCallThreshold": "printWith"
}
```

**Comment by duailibe** [comment]
@lehni We prevent adding any options if we can: https://prettier.io/docs/en/option-philosophy.html

**Comment by duailibe** [comment]
Making it configurable won't solve the problem because at some places you may want 3, others 4, others respect printWidth.

**Comment by apazzolini** [comment]
I still hold my viewpoint from #1565 that Prettier should do for function chaining exactly what it does for objects literals: if the user wrote it on one line and it fits, leave it. If it doesn't fit on one line or the user added a line-break, split out the chained calls.

It's much less arbitrary than a magic number (3), doesn't introduce a configuration option, and I can't see any scenario it'll introduce undesirable or unpredictable behavior in.

**Comment by thorn0** [comment]
Calls without arguments probably can be not counted in this condition:
```js
callExpressionCount >= 3
```
So it'd become 
```js
argumentfulCallExpressionCount >= 3
```

**Comment by lehni** [comment]
I completely agree with @apazzolini here, this is a great proposal, and would solve the awkwardness resulting from hard-coding a completely arbitrary threshold. @duailibe curious to hear your thoughts on this.

**Comment by shellscape** [comment]
I agree with @apazzolini as well. It'd bonkers that this issue has been floating around for a year now. I've gone so far as to use a patching module to hot-patch prettier to modify this behavior.

Come on folks! @apazzolini has an extremely reasonable proposal. Let's see it happen!

**Comment by j-f1** [comment]
My PR: #4765. Should I get that up to date and merge it @prettier/core?

**Comment by Webman** [comment]
How to use this feature? Is there any options, settings?

**Comment by epidemian** [comment]
I'm pretty sure this has already been fixed. The examples given here now [don't add excessive line breaks](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEAHAhgGwEsoATAJwgFs5MBeTDMugPk2AB0pNNIoMG6DGGQB0MCABkIAdzhkAwrjRwAFAEoRZONny4wqgPQBtduwDqAfQC6BgOYAaTAHInagNyduWmAFcyXNDpaejQRNB1CGBUXDS0ANzlldREAKwhiaNcPKABfbM44AA9sODAo-Xx8UNwogEZYhBI5ZPhCqLUNcQAhVSciKABrTFaYWtcC4tLyuErqqIAmBtJmzqL2zoge6P6hkfnxqCKSspUKqpEalQBmJaayFrX1Da2+4l21q4OjqdOZ88uACy3FZiR4dMSbXo7YZrAGuED2EAQbAwQjoZCgXBkCjSAAKWIQaGQIAI0lwAE8iYiAEZkPQDOAwADKeDAxFsyGEPjgiLglGpcBITRIElwUFsPlwtjgADEIGRKDVUeLibgfOIESAABYwSj4MxayJwcJ6OBMwmRQhxSLk4lgNBUkDEZRkGC4um2RXIABmBGUiJSaEKXTpYAZzNw1AkxDgPr9PJAgcKTPZ+DgAEUfBB4HGqgm8GQXcTqbgBfhNdgyMQYGZCCQYFrkAAOAAMiMrEGUZjp2GJleNcgSmoAjln4O7kUSUEoALSHQWCzVaUeELTuqVepC+vOI5SUQicsjc3epjNj2Nb+OImCl2v1xtIebXumEfq2eRUTcgY0AVk1PmUAAVUsp23f0QDibkAElllgJkwCrFEAEFSCZGByTTXNlByHIgA) :)
